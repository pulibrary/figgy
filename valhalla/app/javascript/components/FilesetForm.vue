<template>
  <div>
    <h2>Edit <small>the selected item</small></h2>
    <div class="row">
      <div class="col-md-12">
        <form class="formContent form-inline">
          <div class="form-group">
            <label class="control-label" for="label">Label</label>
            <input @input="updateSingle()" v-model="singleForm.label" type="text" name="label" id="label" value="1" class="form-control">
          </div>
          <div class="form-group">
            <label class="control-label" for="pageType">Page Type</label>
            <select @change="updateSingle()" v-model="singleForm.pageType" id="pageType" class="form-control">
              <option value="single">Single Page (Default)</option>
              <option value="non-paged">Non-Paged</option>
              <option value="facing">Facing Pages</option>
            </select>
          </div>
          <br/>
          <div class="form-group">
            <div class="checkbox">
              <label>
                <input @change="updateThumbnail()" v-model="isThumbnail" id="isThumbnail" type="checkbox" :value="thumbnail">
                Set as Thumbnail</a>
              </label>
            </div>
            <div class="checkbox">
              <label>
                <input @change="updateStartPage()" v-model="isStartPage" id="isStartPage" type="checkbox" :value="startPage">
                Set as Start Page</a>
              </label>
            </div>
          </div>
          <input id="canvas_id" type="hidden" name="canvas_id">
        </form>
      </div>
      <div class="col-md-12">
          <a :href="singleForm.editLink" id="replace-file-button" class="btn btn-default btn-lg">Replace or Delete File</a>
      </div>
    </div>
  </div>
</template>

<script>
export default {
  name: 'fileset-form',
  computed: {
    thumbnail: function () {
      return this.$store.state.thumbnail
    },
    startPage: function () {
      return this.$store.state.startPage
    },
    isStartPage () {
      var id = this.$store.state.selected[0].id
      return this.$store.state.startPage === id
    },
    isThumbnail () {
      var id = this.$store.state.selected[0].id
      return this.$store.state.thumbnail === id
    },
    singleForm () {
      return {
        label: this.$store.state.selected[0].label,
        id: this.$store.state.selected[0].id,
        pageType: this.$store.state.selected[0].pageType,
        url: this.$store.state.selected[0].url,
        editLink: '/catalog/parent/' + this.$store.state.id + '/' + this.$store.state.selected[0].id
      }
    }
  },
  methods: {
    updateStartPage () {
      var startPage = this.$store.state.selected[0].id
      this.$store.dispatch('updateStartPage', startPage)
    },
    updateThumbnail () {
      var thumbnail = this.$store.state.selected[0].id
      this.$store.dispatch('updateThumbnail', thumbnail)
    },
    updateSingle () {
      var changeList = this.$store.state.changeList
      var images = this.$store.state.images
      var index = this.$store.state.images.map(function (img) {
        return img.id
      }).indexOf(this.$store.state.selected[0].id)
      images[index] = this.singleForm

      if (changeList.indexOf(this.$store.state.selected[0].id) === -1) {
        changeList.push(this.$store.state.selected[0].id)
      }

      this.$store.dispatch('updateChanges', changeList)
      this.$store.dispatch('updateImages', images)
    }
  }
}
</script>

<!-- Add "scoped" attribute to limit CSS to this component only -->
<style scoped>

.form-inline fieldset {
  display: inline-block;
}

.form-inline .row {
  margin-top:15px;
}

#singleSelected .form-group {
  line-height: 3em;
}

.checkbox {
  display: block;
}

#replace-file-button {
  margin-bottom: 10px;
}

</style>
