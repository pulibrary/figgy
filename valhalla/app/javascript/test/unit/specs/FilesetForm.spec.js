import Vuex from 'vuex'
import { mount, createLocalVue } from 'vue-test-utils'
import FilesetForm from '@/components/FilesetForm'
import Fixtures from '@/test/fixtures/image-collection'
const localVue = createLocalVue()
localVue.use(Vuex)

describe('FilesetForm.vue', () => {
  let wrapper
  let options
  let actions
  let state
  let store

  beforeEach(() => {
    actions = {
      updateStartPage: jest.fn(),
      updateThumbnail: jest.fn(),
      updateChanges: jest.fn(),
      updateImages: jest.fn()
    }
    state = {
      images: Fixtures.imageCollection,
      selected: Fixtures.selected,
      changeList: Fixtures.changeList,
      thumbnail: Fixtures.thumbnail,
      startPage: Fixtures.startPage
    }
    store = new Vuex.Store({
      state,
      actions
    })

    options = {
      computed: {
        thumbnail: function () {
          return state.thumbnail
        },
        startPage: function () {
          return state.startPage
        },
        isStartPage () {
          var id = state.selected[0].id
          return state.startPage === id
        },
        isThumbnail () {
          var id = state.selected[0].id
          return state.thumbnail === id
        },
        singleForm () {
          return {
            label: state.selected[0].label,
            id: state.selected[0].id,
            pageType: state.selected[0].pageType,
            url: state.selected[0].url,
            editLink: '/catalog/parent/' + state.id + '/' + state.selected[0].id
          }
        }
      }
    }
  })

  it('allows the selection of a thumbnail image', () => {
    const wrapper = mount(FilesetForm, { options, store, localVue })
    wrapper.find('#isThumbnail').trigger('change')
    expect(actions.updateThumbnail).toHaveBeenCalled()
  })

  it('allows the selection of a start page', () => {
    const wrapper = mount(FilesetForm, { options, store, localVue })
    wrapper.find('#isStartPage').trigger('change')
    expect(actions.updateStartPage).toHaveBeenCalled()
  })

  it('allows the selection of a page type', () => {
    const wrapper = mount(FilesetForm, { options, store, localVue })
    wrapper.find('#pageType').trigger('change')
    expect(actions.updateChanges).toHaveBeenCalled()
    expect(actions.updateImages).toHaveBeenCalled()
  })

  it('allows the input of a label', () => {
    const wrapper = mount(FilesetForm, { options, store, localVue })
    wrapper.find('#label').trigger('input')
    expect(actions.updateChanges).toHaveBeenCalled()
    expect(actions.updateImages).toHaveBeenCalled()
  })

  it('has the expected html structure', () => {
    const wrapper = mount(FilesetForm, { options, store, localVue })
    expect(wrapper.element).toMatchSnapshot()
  })

})
