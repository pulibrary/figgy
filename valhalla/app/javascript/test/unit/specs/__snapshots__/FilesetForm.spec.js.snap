// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FilesetForm.vue has the expected html structure 1`] = `
<div>
  <h2>
    Edit 
    <small>
      the selected item
    </small>
  </h2>
   
  <div
    class="row"
  >
    <div
      class="col-md-12"
    >
      <form
        class="formContent form-inline"
      >
        <div
          class="form-group"
        >
          <label
            class="control-label"
            for="label"
          >
            Label
          </label>
           
          <input
            class="form-control"
            id="label"
            name="label"
            type="text"
            value="1"
          />
        </div>
         
        <div
          class="form-group"
        >
          <label
            class="control-label"
            for="pageType"
          >
            Page Type
          </label>
           
          <select
            class="form-control"
            id="pageType"
          >
            <option
              value="single"
            >
              Single Page (Default)
            </option>
             
            <option
              value="non-paged"
            >
              Non-Paged
            </option>
             
            <option
              value="facing"
            >
              Facing Pages
            </option>
          </select>
        </div>
         
        <br />
         
        <div
          class="form-group"
        >
          <div
            class="checkbox"
          >
            <label>
              <input
                id="isThumbnail"
                type="checkbox"
              />
              
              Set as Thumbnail
            </label>
          </div>
           
          <div
            class="checkbox"
          >
            <label>
              <input
                id="isStartPage"
                type="checkbox"
              />
              
              Set as Start Page
            </label>
          </div>
        </div>
         
        <input
          id="canvas_id"
          name="canvas_id"
          type="hidden"
        />
      </form>
    </div>
     
    <div
      class="col-md-12"
    >
      <a
        class="btn btn-default btn-lg"
        href="/catalog/parent/undefined/50b5e49b-ade7-4278-8265-4f72081f26a5"
        id="replace-file-button"
      >
        Replace or Delete File
      </a>
    </div>
  </div>
</div>
`;
